File: src/09.tree/44.538.ts
import { TreeNode } from "./01.100";

export function convertBST(root: TreeNode | null): TreeNode | null {
    let start = 0;
    const dfs = (node: TreeNode | null) => {
        if (!node) {
            return;
        }

        dfs(node.right);
        node.val += start;
        start = node.val;
        dfs(node.left);
    };
    dfs(root);
    return root;
}

File: src/09.tree/44.538.go
package tree

func ConvertBST(root *TreeNode) *TreeNode {
	var dfs func(node *TreeNode)
	pre := 0
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		dfs(node.Right)
		node.Val += pre
		pre = node.Val
		dfs(node.Left)
	}
	dfs(root)
	return root
}
